File: core/src/nl/t64/game/rpg/screens/world/MapSubject.java
package nl.t64.game.rpg.screens.world;

import com.badlogic.gdx.graphics.Color;

import java.util.ArrayList;
import java.util.List;


public class MapSubject {

    private final List<MapObserver> observers = new ArrayList<>();

    public void addObserver(MapObserver observer) {
        observers.add(observer);
    }

    public void removeObserver(MapObserver observer) {
        observers.remove(observer);
    }

    public void removeAllObservers() {
        observers.clear();
    }

    public void notifyMapWillChange(Runnable changeMap, Color transitionColor) {
        observers.forEach(mapObserver -> mapObserver.onNotifyMapWillChange(changeMap, transitionColor));
    }

    public void notifyMapChanged(GameMap currentMap) {
        observers.forEach(mapObserver -> mapObserver.onNotifyMapChanged(currentMap));
    }

}

File: core/src/nl/t64/game/rpg/subjects/MapSubject.kt
package nl.t64.game.rpg.subjects

import com.badlogic.gdx.graphics.Color
import nl.t64.game.rpg.screens.world.GameMap


class MapSubject {

    private val observers: MutableList<MapObserver> = ArrayList()

    fun addObserver(observer: MapObserver) {
        observers.add(observer)
    }

    fun removeObserver(observer: MapObserver) {
        observers.remove(observer)
    }

    fun removeAllObservers() {
        observers.clear()
    }

    fun notifyFadeOut(actionAfterFade: () -> Unit, transitionColor: Color) {
        observers.forEach { it.onNotifyFadeOut(actionAfterFade, transitionColor) }
    }

    fun notifyMapChanged(currentMap: GameMap) {
        observers.forEach { it.onNotifyMapChanged(currentMap) }
    }

    fun notifyShakeCamera() {
        observers.forEach { it.onNotifyShakeCamera() }
    }

    fun notifyStartCutscene(cutsceneId: String) {
        observers.forEach { it.onNotifyStartCutscene(cutsceneId) }
    }

}
